### DISCUSSION TITLE
Arduino Due - Is a Control board required?

### DISCUSSION THREAD

**Discussion by roberalbe** [grblHAL/Controllers#14] [Q&A]
Hi,

I'm new to DIY CNC stepper systems (Noob).  I bought an old Shop Sabre 4' x 4' CNC router table that was computer controlled but the computer system was bad.  Through some research I decided to switch to a grblhal setup with a Arduino Due.  The table has large NEMA 34 steppers with standalone Stepper Driver boards that just need step/dir input.  I assumed then all I needed was to hookup the step/dir outputs direct off the Arduino Due to the driver boards.  I used the web configurator and configured the due with a Ramps 1.6 board as I need a ganged dual motor Y-axis.  Testing the X-axis I get motion but it has erratic, jittery speed from point to point and not smooth.  However it seems to be repeatable going back and forth from point to point.  In testing I narrowed it down to the step/dir interface.  So a couple questions.

1) Is the Ramps 1.6 board required?  Should have pulled the step/dir from that board and not the Due directly?

2) Is there a better control board setup that would be better for this application?

3) Is there some settings that I'm missing that may smooth this out?

Thanks
Bob

P.S.  used a couple different Gsenders but most of the testing has been with UGS.

**Comment by terjeio** [comment]
The stepper driver board may not work with the 3.3V signals the Due provides, if so you will have to add level shifters.

1. No, and pulling the step/dir signals from it does not change anything.
2. There are grblHAL boards that provides buffered 5V step/dir signals which may be better suited. Do you have access to the driver board documentation?
3. Perhaps. Inverting the step signals could be one, increasing step pulse length another. Check the driver documentation if available.

**Reply by roberalbe** [reply]
Terje,

Sorry I forgot to mention I have an optocoupler board between the Due and Drive to handle the voltage difference(See Below).  Yes I do have the documentation on the drives, and surprisingly good documentation.  The drive is configured for a 2MHz max frequency.
1) was the optocoupler board a bad choice? I can switch to a level shifter.

![image](https://github.com/user-attachments/assets/2b8740c9-2553-43bd-85f7-2b19edd6d6a9)

![image](https://github.com/user-attachments/assets/cd344566-14bb-4380
-910a-
d4607ba27a00)


https://github.com/user-attachments/assets/7cb909b0-cd24-4530-a0a8-fa3b244200be

**Reply by roberalbe** [reply]
![image](https://github.com/user-attachments/assets/4bf09836-748a-497b-a901-20cf5611d13e)

**Reply by terjeio** [reply]
The optocoupler board seems to be an older version and might be a bit slow. Was it used in the original setup?

> Testing the X-axis I get motion but it has erratic, jittery speed from point to point and not smooth.

You should set $29 (pulse delay) to at least 3 to be in compliance with _Dir Electrical Specs_. This does not affect motion from point to point without direction changes though.

If going back and forth a few times do you lose position due to lost steps?

**Reply by terjeio** [reply]
BTW what is the series resistor value on the input side?
There is also a marking in the upper right corner of the board that suggests there are different input voltage settings somewhere, or that you have to select a resistors package that corresponds to the input.

**Reply by roberalbe** [reply]
-"**The optocoupler board seems to be an older version and might be a bit slow. Was it used in the original setup**?"
No this wasn't in the original setup, I had it already laying around and tried it.  I'm thinking of just getting a level shifter.

-"**You should set $29 (pulse delay) to at least 3 to be in compliance with Dir Electrical Specs. This does not affect motion from point to point without direction changes though.**"
I did try adjusting pulse delay and pulse length in both directions, but it didn't change the motion symptoms at all

-"**If going back and forth a few times do you lose position due to lost steps?**"
No, it seems to start and stop very repeatable even after dozens of moves back and forth

-"**BTW what is the series resistor value on the input side?
There is also a marking in the upper right corner of the board that suggests there are different input voltage settings somewhere, or that you have to select a resistors package that corresponds to the input.**"
Yeah, I was confused by that as well.  there isn't any configuration of this board or documentation.  That's another reason to switch to level shifter.  I don't think it is the issue however.  I did hook the drive up direct as just to try, as the drive documentation says it can go as low as 4vdc, and may be in the margin of error and it seemed to work but with the exact same jittery motion.  (Jitter may be the wrong decription, it seems to stutter every 1second or so, and seems to be consistent in timing and you may be able to see it in the video on the bottom LED).

Thanks for all the input!

**Reply by roberalbe** [reply]
Well, did a bunch more tinkering and now I'm thinking the Due board is bad.  I had a Arduino MEGA and threw it on with GRBL 1.1 and then the X axis runs buttery smooth! 
 
If I have to replace the driver board, what driver board would you recommend for the application?

**Reply by terjeio** [reply]
The Arduino MEGA is a 5V device, the Due is 3.3V. If the resistors in series with the optos are too large for 3.3V the optos may degrade the stepper pulse quite a lot.
It looks like the resistor array is socketed, did the board come with a number of different values to choose from?
